test/Tests.Common/TestTimeProvider.cs

```diff
@@ -0,0 +1,151 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace Yarp.Tests.Common;
+
+/// <summary>
+/// Simulates passage of time, used for testing.
+/// </summary>
+/// <remarks>
+/// This timer doesn't track real time, but instead tracks virtual time.
+/// Time only advances when any of the following methods are called:
+/// <list type="bullet">
+/// <item><see cref="Advance"/></item>
+/// <item><see cref="AdvanceTo(TimeSpan)"/></item>
+/// </list>
+/// </remarks>
+public class TestTimeProvider : TimeProvider
+{
+    private readonly List<TestTimer> _timers = new();
+
+    private TimeSpan _currentTime;
+
+    public int TimerCount => _timers.Count;
+
+    public override long TimestampFrequency => TimeSpan.TicksPerSecond;
```

Comment:
```suggestion
    public override long TimestampFrequency => TimeSpan.TicksPerSecond * 7;
```

Nit: Can we have this return a non-standard value, so that we can catch product errors that our tests would be happy with (e.g. we offset a timestamp by `timeSpan.Ticks` somewhere instead of going through `TimeProvider`)?
`TimestampFrequency` and `GetTimestamp` are the only places that should have to change.